Adding override code to create RestAPI Lambda Authorizers (#5614)
aws-amplify/docs: src/pages/cli/restapi/override.mdx

endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
the identity of the user for accessing other services like AWS AppSync or S3.

## Authorize API requests with Lambda Authorizer

> In a similar way like Cognito Authorizer, Amplify CLI also does not support Lambda authorizers out-of-box. To implement this functionality, you must override your REST API and add a Lambda authorizer yourself by adding the following code into the `override(...)` function. The following steps are walk-through to create Token-Based Lambda Authorizer.

Reviewer: ```suggestion
While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality my overriding your REST API resources. The following steps are walk-through to create token-based Lambda authorizer.
```

Author: Hello Rene,
I have updated introduction as suggested.

> First, assuming the Lambda Authorizer function you would like to use as an authorizer is configured with your Amplify Project, create a parameter that resolves to Lambda Function ARN:

Reviewer: I'd rewrite this to also link to the flow where you run `amplify add function` and request customers to review there.

Reviewer: Can you illustrate where I create this parameter? For example in xyz file... or "after you run...".

Author: Added, steps to create lambda function and mentioned file name to add the code

Make sure to replace `<your-api-resource-name>` with the name of your Rest API resource.
This is the name of the folder in `amplify/backend/api` that was created when
you added an Rest API resource to your Amplify project.

</Callout>

Finally, update the security methods for all of the paths in your REST API to use this new Lambda authorizer. You can also add the Authorization header as a parameter on incoming requests for these paths as a place for users to provide their Auth token. 

```ts
for (const path in resources.restApi.body.paths) {
    // Add the Authorization header as a parameter to requests
    resources.restApi.addPropertyOverride(
      `Body.paths.${path}.x-amazon-apigateway-any-method.parameters`,
      [
        ...resources.restApi.body.paths[path]["x-amazon-apigateway-any-method"]
          .parameters,
        {
          name: "Authorization",
          in: "header",
          required: false,
          type: "string",
        },
      ]
    );
    // Use your new Lambda authorizer for security
    resources.restApi.addPropertyOverride(
      `Body.paths.${path}.x-amazon-apigateway-any-method.security`,
      [ { Lambda: [], }, ]
    );
  }
```

> You can also configure Request Based Lambda Authorizer referring the OpenAPI extension, [`x-amazon-apigateway-authorizer`](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-authorizer.html).

Reviewer: ```suggestion
You can also configure request-based Lambda authorizers referenced in the OpenAPI extension, [`x-amazon-apigateway-authorizer`](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-authorizer.html).
```

> Additionally you can add more advanced logic to only use the Lambda authorizer

Reviewer: I'd suggest adding a code sample here or exclude it for readability.

with some paths or methods.

> When performing requests to your REST API, make sure to add the Authorization header with an token required by Lambda Authorizer Function.

Reviewer: Show a code snippet here with for the client-side execution

Author: Hello,
Referring the documentation for cognito authorizer, I could see client code is mentioned in lib section. Hence added the client code for lambda authorizer in lib section -> Define authorization rules.

endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
the identity of the user for accessing other services like AWS AppSync or S3.

## Authorize API requests with Lambda authorizer

> While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps are walk-through to create token-based Lambda authorizer.

Reviewer 2: ```suggestion
While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps will walk you through how to create token-based Lambda authorizer.
```

> Firstly, you need to have a Lambda authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`

Reviewer 2: ```suggestion
First, you need to have a Lambda authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your Amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`
```

> Firstly, you need to have an  Lambda Authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`

Reviewer: ```suggestion
Firstly, you need to have a Lambda authorizer function with required [authorization logic](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html#api-gateway-lambda-authorizer-lambda-function-create) in your amplify project to use it as an authorizer. Refer to the steps to [set up a function](https://docs.amplify.aws/cli/function/) using `amplify add function`
```

After running `amplify override api`, add the following code to `override(...)` function.
Initially, create a parameter that resolves to Lambda Function ARN

```ts
// Replace the following with your Function resource name
const functionResourcename = "<your-function-resource-name>";
const functionArnParameter = "FunctionArn";

//  Adding parameter to your Cloud Formation Template for Authorizer function arn

>   resources.addCfnParameter({
>     type: "String",
>     description: "The ARN of an existing Lambda Function to authorize requests",
>     default: "NONE",
>   },
>   functionArnParameter,
>   { "Fn::GetAtt": [`function${functionResourcename}`, "Outputs.Arn"], }
> );

Reviewer 2: ```suggestion
resources.addCfnParameter(
  {
    type: 'String',
    description: 'The ARN of an existing Lambda Function to authorize requests',
    default: 'NONE',
  },
  functionArnParameter,
  { 'Fn::GetAtt': [`function${functionResourcename}`, 'Outputs.Arn'] }
)
```

nit for formatting

```

<Callout warning>

Make sure to replace `<your-function-resource-name>` with the name of your function resource.
This is the name of the folder in `amplify/backend/function` that was created when
you added an function resource to your Amplify project.

</Callout>

Next, define the Lambda authorizer using the OpenAPI extension, [`x-amazon-apigateway-authorizer`](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-swagger-extensions-authorizer.html). This change will be applied by modifying the security definition of your REST API:

```ts

>  // Create the authorizer using the functionArnParameter parameter defined above
>  resources.restApi.addPropertyOverride("Body.securityDefinitions", {
>     Lambda: {
>       type: "apiKey",
>       name: "Authorization",
>       in: "header",
>       "x-amazon-apigateway-authtype": "oauth2",
>       "x-amazon-apigateway-authorizer": {
>         type: "token",
>         authorizerUri: 
>         {
>           'Fn::Join': ['', ["arn:aws:apigateway:",{ Ref: 'AWS::Region' },":lambda:path/2015-03-31/functions/",{ Ref: functionArnParameter },"/invocations"]],
>         },
>         authorizerResultTtlInSeconds: 0
>       },
>     },
>   });

Reviewer 2: ```suggestion
 // Create the authorizer using the functionArnParameter parameter defined above
resources.restApi.addPropertyOverride('Body.securityDefinitions', {
  Lambda: {
    type: 'apiKey',
    name: 'Authorization',
    in: 'header',
    'x-amazon-apigateway-authtype': 'oauth2',
    'x-amazon-apigateway-authorizer': {
      type: 'token',
      authorizerUri: {
        'Fn::Join': [
          '',
          [
            'arn:aws:apigateway:',
            { Ref: 'AWS::Region' },
            ':lambda:path/2015-03-31/functions/',
            { Ref: functionArnParameter },
            '/invocations',
          ],
        ],
      },
      authorizerResultTtlInSeconds: 0,
    },
  },
})
```

```

As API Gateway needs permission to invoke the Authorizer lambda function, add resource based policy to the function using following code:

```ts

> // Adding Resource Based policy to Lambda Authorizer function

Reviewer: ```suggestion
// Adding Resource Based policy to Lambda authorizer function
```

> // Adding Resource Based policy to Lambda authorizer function
> resources.addCfnResource(
>     {
>       type: "AWS::Lambda::Permission",
>       properties: {
>             Action: "lambda:InvokeFunction",
>             FunctionName: {Ref: functionArnParameter},
>             Principal: "apigateway.amazonaws.com",
>             SourceArn:{
>                 "Fn::Join": [
>                   "",
>                   [
>                     "arn:aws:execute-api:",
>                     {
>                       "Ref": "AWS::Region"
>                     },
>                     ":",
>                     {
>                       "Ref": "AWS::AccountId"
>                     },
>                     ":",
>                     {
>                       "Ref": "<your-API-resource-name>"
>                     },
>                     "/*/*"
>                   ]
>                 ]
>               }
>             }
>     },"LambdaAuthorizerResourceBasedPolicy");
> 

Reviewer 2: ```suggestion
// Adding Resource Based policy to Lambda authorizer function
resources.addCfnResource(
  {
    type: 'AWS::Lambda::Permission',
    properties: {
      Action: 'lambda:InvokeFunction',
      FunctionName: { Ref: functionArnParameter },
      Principal: 'apigateway.amazonaws.com',
      SourceArn: {
        'Fn::Join': [
          '',
          [
            'arn:aws:execute-api:',
            {
              Ref: 'AWS::Region',
            },
            ':',
            {
              Ref: 'AWS::AccountId',
            },
            ':',
            {
              Ref: '<your-API-resource-name>',
            },
            '/*/*',
          ],
        ],
      },
    },
  },
  'LambdaAuthorizerResourceBasedPolicy'
)
```

```
<Callout warning>

> Make sure to replace `<your-api-resource-name>` with the name of your Rest API resource.
> This is the name of the folder in `amplify/backend/api` that was created when
> you added an Rest API resource to your Amplify project.

Reviewer 2: ```suggestion
Make sure to replace `<your-api-resource-name>` with the name of your Rest API resource. This is the name of the folder in `amplify/backend/api` that was created when you added an Rest API resource to your Amplify project.
```

</Callout>

Finally, update the security methods for all of the paths in your REST API to use this new Lambda authorizer. You can also add the Authorization header as a parameter on incoming requests for these paths as a place for users to provide their Auth token. 

```ts

> for (const path in resources.restApi.body.paths) {
>     // Add the Authorization header as a parameter to requests
>     resources.restApi.addPropertyOverride(
>       `Body.paths.${path}.x-amazon-apigateway-any-method.parameters`,
>       [
>         ...resources.restApi.body.paths[path]["x-amazon-apigateway-any-method"]
>           .parameters,
>         {
>           name: "Authorization",
>           in: "header",
>           required: false,
>           type: "string",
>         },
>       ]
>     );
>     // Use your new Lambda authorizer for security
>     resources.restApi.addPropertyOverride(
>       `Body.paths.${path}.x-amazon-apigateway-any-method.security`,
>       [ { Lambda: [], }, ]
>     );
>   }

Reviewer 2: ```suggestion
for (const path in resources.restApi.body.paths) {
  // Add the Authorization header as a parameter to requests
  resources.restApi.addPropertyOverride(
    `Body.paths.${path}.x-amazon-apigateway-any-method.parameters`,
    [
      ...resources.restApi.body.paths[path]['x-amazon-apigateway-any-method']
        .parameters,
      {
        name: 'Authorization',
        in: 'header',
        required: false,
        type: 'string',
      },
    ]
  )
  // Use your new Lambda authorizer for security
  resources.restApi.addPropertyOverride(
    `Body.paths.${path}.x-amazon-apigateway-any-method.security`,
    [{ Lambda: [] }]
  )
}
```

```
<Callout>

> Note that you can add more advanced logic to only use the Lambda authorizer
> with some paths or methods.

Reviewer 2: ```suggestion
Note that you can add more advanced logic to only use the Lambda authorizer with some paths or methods.
```

</Callout>

> When performing requests to your REST API, make sure to add the Authorization header with an token required by Lambda Authorizer Function. 

Reviewer: ```suggestion
When performing requests to your REST API, make sure to add the Authorization header with an token required by Lambda authorizer function. 
```

endpoint, and you can reuse the verified ID Token in your endpoint resolvers to assume
the identity of the user for accessing other services like AWS AppSync or S3.

## Authorize API requests with Lambda authorizer

> While Amplify CLI also does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps will walk you through how to create token-based Lambda authorizer.

Reviewer: ```suggestion
While Amplify CLI does not support Lambda authorizers natively out-of-box, you can implement this functionality by overriding your REST API resources. The following steps will walk you through how to create token-based Lambda authorizer.
```
